crates/swc_css_minifier/src/compressor/calc_sum.rs

```diff
@@ -1,835 +1,807 @@
 use std::collections::HashMap;
 
-use swc_atoms::JsWord;
+use string_cache::Atom;
+use swc_atoms::{js_word, JsWord, JsWordStaticSet};
+use swc_common::Span;
 use swc_css_ast::*;
 
-use super::{unit::*, Compressor};
+use super::Compressor;
 use crate::compressor::math::{is_calc_function_name, transform_calc_value_into_component_value};
 
-// transform "(simple calc-value)" into "simple calc-value"
-fn remove_unnecessary_nesting_from_calc_sum(calc_sum: &mut CalcSum) {
-    if calc_sum.expressions.len() == 1 {
-        match &calc_sum.expressions[0] {
-            CalcProductOrOperator::Product(CalcProduct {
-                expressions: calc_product_expressions,
-                ..
-            }) if calc_product_expressions.len() == 1 => {
-                if let CalcValueOrOperator::Value(CalcValue::Sum(CalcSum {
-                    expressions: nested_expressions,
-                    span: nested_span,
-                })) = &calc_product_expressions[0]
-                {
-                    calc_sum.span = *nested_span;
-                    calc_sum.expressions = nested_expressions.to_vec();
+// Calc-operator node of the calculation tree
+// https://www.w3.org/TR/css-values-4/#calculation-tree-calc-operator-nodes
+#[derive(Debug, Clone)]
+enum CalcOperatorNode {
+    Sum(Vec<CalcNode>),
+    Product(Vec<CalcNode>),
+    Negate(CalcNode),
+    Invert(CalcNode),
+}
+
+#[derive(Debug, Clone)]
+enum CalcNode {
+    Number(Number),
+    Dimension(Dimension),
+    Percentage(Percentage),
+    Constant(Ident),
+    Function(Function),
+    OperatorNode(Box<CalcOperatorNode>),
+}
+
+fn get_precision(n: f64) -> u32 {
+    let n_as_str = n.to_string();
+    n_as_str
+        .find('.')
+        .map_or(0, |sep| (n_as_str.len() - sep) as u32 - 1)
+}
```

Comment:
Why do not use `.fract()`?

Reply:
`1.4.fract()` returns "0.33333333", but we need "1" instead to have the "right" precision. The rust `to_string()` function seems clever for this purpose.